src/pounder/mod.rs

```diff
@@ -32,34 +32,109 @@ const ATT_LE3_PIN: u8 = 8 + 3;
+        // the data encoded into 4-bit QSPI data so that IO3 can be driven low.
+        match self.mode {
+            ad9959::Mode::SingleBitTwoWire => {
+                // Encode the data into a 4-bit QSPI pattern.
+
+                // In 4-bit mode, we can send 2 bits of address and data per byte transfer. As
+                // such, we need at least 4x more bytes than the length of data. To avoid dynamic
+                // allocation, we assume the maximum transaction length for single-bit-two-wire is
+                // 2 bytes.
+                let mut encoded_data: [u8; 12] = [0; 12];
+
+                if (data.len() * 4) > (encoded_data.len() - 4) {
+                    return Err(Error::Bounds);
+                }
+
+                // Encode the address into the first 4 bytes.
+                for address_bit in 0..8 {
+                    let offset: u8 = {
+                        if address_bit % 2 == 0 {
+                            4
+                        } else {
+                            0
+                        }
+                    };
+
+                    if addr & address_bit != 0 {
+                        encoded_data[(address_bit >> 1) as usize] |= 1 << offset;
+                    }
+                }
+
+                // Encode the data into the remaining bytes.
+                for byte_index in 0..data.len() {
+                    let byte = data[byte_index];
+                    for address_bit in 0..8 {
+                        let offset: u8 = {
+                            if address_bit % 2 == 0 {
+                                4
+                            } else {
+                                0
+                            }
+                        };
+
+                        if byte & address_bit != 0 {
+                            encoded_data[(byte_index + 1) * 4 + (address_bit >> 1) as usize] |= 1 << offset;
+                        }
+                    }
+                }
+
+                let (encoded_address, encoded_payload) = {
+                    let end_index = (1 + data.len()) * 4;
+                    (encoded_data[0], &encoded_data[1..end_index])
+                };
+
+                self.qspi.write(encoded_address, &encoded_payload).map_err(|_| Error::Qspi)
+            },
+            ad9959::Mode::FourBitSerial => {
+                self.qspi.write(addr, &data).map_err(|_| Error::Qspi)
+            },
+            _ => {
+                Err(Error::Qspi)
+            }
+        }
     }
 
     fn read(&mut self, addr: u8, mut dest: &mut [u8]) -> Result<(), Error> {
         if (addr & 0x80) != 0 {
             return Err(Error::InvalidAddress);
         }
+
+        // It is not possible to read data from the AD9959 in single bit two wire mode because the
+        // QSPI interface assumes that data is always received on IO1.
+        if self.mode == ad9959::Mode::SingleBitTwoWire {
```

Comment:
I think the only mode that works for reads is `FourBitSerial`.

```suggestion
        if self.mode != ad9959::Mode::FourBitSerial {
```